usaspending_api/common/api_request_utils.py

```diff
@@ -171,9 +188,84 @@ def validate_post_request(self, request):
                         if filt['operation'] == 'range':
                             if not isinstance(filt['value'], list) or len(filt['value']) != 2:
                                 raise Exception("Invalid value, operation 'range' requires an array value of length 2")
+                        if filt['operation'] == 'range_intersect':
+                            if not isinstance(filt['field'], list) or len(filt['field']) != 2:
+                                raise Exception("Invalid field, operation 'range_intersect' requires an array of length 2 for field")
+                            if (not isinstance(filt['value'], list) or len(filt['value']) != 2) and 'value_format' not in filt:
+                                raise Exception("Invalid value, operation 'range_intersect' requires an array value of length 2, or a single value with value_format set to a ranged format (such as fy)")
                     else:
                         raise Exception("Malformed filter - missing field, operation, or value")
 
+    # Special operation functions follow
+
+    # Range intersect function - evaluates if a range defined by two fields overlaps
+    # a range of values
+    # Here's a picture:
+    #                 f1 - - - f2
+    #                       r1 - - - r2     - Case 1
+    #             r1 - - - r2               - Case 2
+    #                 r1 - - - r2           - Case 3
+    # All of the ranges defined by [r1,r2] intersect [f1,f2]
+    # i.e. f1 <= r2 && r1 <= f2 we intersect!
+    # Returns: Q object to perform this operation
+    # Parameters - Make sure these are in order:
+    #           fields - A list defining the fields forming the first range (in order)
+    #           values - A list of the values which define the second range (in order)
+    def range_intersect(self, fields, values):
+        # Create the Q filter case
+        q_case = {}
+        q_case[fields[0] + "__lte"] = values[1]  # f1 <= r2
+        q_case[fields[1] + "__gte"] = values[0]  # f2 >= r1
+        return Q(**q_case)
+
+
+# Handles autocomplete requests
+class AutoCompleteHandler():
+    @staticmethod
+    # Data set to be searched for the value, and which fields to look in
+    # Mode is either "contains" or "startswith"
+    def get_values_and_counts(data_set, fields, value, mode="contains"):
+        value_dict = {}
+        count_dict = {}
+
+        if mode == "contains":
+            mode = "__icontains"
+        elif mode == "startswith":
+            mode = "__istartswith"
+
+        for field in fields:
+            q_args = {}
+            q_args[field + mode] = value
+            value_dict[field] = list(set(data_set.filter(Q(**q_args)).values_list(field, flat=True)))  # Why this weirdness? To ensure we eliminate duplicates
```

Comment:
It can, and we already support full text search; but I do not think that it supports functionality such as 'startswith'; although I am not sure about the usefulness of that in general. If you are ok with losing that ability I can move it to using the already implemented search functionality.